File: Javascript/739. Daily Temperatures.js
/**
 * Link: https://leetcode.com/problems/daily-temperatures/
 * Author: Dung Nguyen Tien (Chris)
 * @param {number[]} T
 * @return {number[]}
 */
var dailyTemperatures = function(T) {
    // monotonic stack
    var st = [];
    var ret = Array(T.length).fill(0);
    
    for (var i = 0; i < T.length; i++) {
        // if the array going down, keep pushing to stack the day
        // if i point to the higher T
        while(st.length > 0 && T[i] > T[st[0]]) {
            // pop the right after lower in stack
            var index = st.shift();
            // calculate the day
            ret[index] = i - index;
            // i still point to the higher
        }
        st.unshift(i);
    }
    
    return ret;
};

File: C#/739. Daily Temperatures.cs
/* 
 * Link: https://leetcode.com/problems/daily-temperatures/
 * Author: Dung Nguyen Tien (Chris)
*/

public class Solution {
    public int[] DailyTemperatures(int[] T) {
        var st = new Stack<int>();
        var map = new Dictionary<int, int>();
        var result = new int[T.Length];
        Array.Fill(result, -1);
        
        for(var i = 0; i < T.Length; i++) {
            var t = T[i];
            
            while(st.Count > 0 && T[st.Peek()] < t) {
                result[st.Pop()] = i;
            }
            
            st.Push(i);
        }
        
        for (var i = 0; i < T.Length; i++) {
            if (result[i] == -1) {
                result[i] = 0;
            }
            else {
                result[i] = result[i] - i;
            }
        }
        
        return result;
    }
}